arrows/vxl/average_frames.cxx

```diff
@@ -0,0 +1,722 @@
+void online_frame_averager<PixType>
+::process_frame( const vil_image_view< PixType >& input,
+                 vil_image_view< PixType >& average,
+                 vil_image_view< double >& variance )
+{
+  // Check if this is the first time we have processed a frame of this size
+  if( dev1_tmp_space_.ni() != input.ni() ||
+      dev1_tmp_space_.nj() != input.nj() ||
+      dev1_tmp_space_.nplanes() != input.nplanes() )
+  {
+    dev1_tmp_space_.set_size( input.ni(), input.nj(), input.nplanes() );
+    dev2_tmp_space_.set_size( input.ni(), input.nj(), input.nplanes() );
+    variance.set_size( input.ni(), input.nj(), input.nplanes() );
+    variance.fill( 0.0 );
+    this->process_frame( input, average );
+    return;
+  }
+
+  // Calculate difference from last average
+  vil_math_image_abs_difference( input, this->last_average_, dev1_tmp_space_ );
+
+  // Update internal average
+  this->process_frame( input, average );
+
+  // Update the variance
+  vil_math_image_abs_difference( input, average, dev2_tmp_space_ );
+  vil_math_image_product( dev1_tmp_space_, dev2_tmp_space_, dev1_tmp_space_ );
+  variance.deep_copy( dev1_tmp_space_ );
+}
+
+template <typename PixType>
+bool online_frame_averager<PixType>
+::has_resolution_changed( const vil_image_view< PixType >& input )
+{
+  return ( input.ni() != this->last_average_.ni() ||
+           input.nj() != this->last_average_.nj() ||
+           input.nplanes() != this->last_average_.nplanes() );
+}
+
+// Helper function to allocate a completely new image, and cast the input image
+// to whatever specified type the output image is, scaling by some factor if set
+// and rounding if enabled, in one pass.
+template <typename inT, typename outT>
+void copy_cast_and_scale( const vil_image_view< inT >& input,
+                          vil_image_view< outT >& output,
+                          bool rounding_enabled,
+                          const inT scale_factor )
+{
+  // Just deep copy if pixel formats equivalent and there is no scale factor
+  if( vil_pixel_format_of(inT()) == vil_pixel_format_of(outT())
+      && scale_factor == 1.0 )
+  {
+    output.deep_copy( input );
+    return;
+  }
+
+  // Determine if any rounding would even be beneficial based on source types
+  if( rounding_enabled )
+  {
+    rounding_enabled = !std::numeric_limits<inT>::is_integer &&
+                       std::numeric_limits<outT>::is_integer;
+  }
+
+  // Resize, cast and copy
+  unsigned ni = input.ni(), nj = input.nj(), np = input.nplanes();
+  output.set_size( ni, nj, np );
+
+  // s* = source properties, d* = dest
+  std::ptrdiff_t sistep=input.istep(), sjstep=input.jstep(), spstep=input.planestep();
+  std::ptrdiff_t distep=output.istep(), djstep=output.jstep(), dpstep=output.planestep();
+
+  // Iterate through image and perform cast and any required ops. We could
+  // shorten this code using vil_transform and functors (binding arguments),
+  // but just to ensure they get inlined...
```

Comment:
This is bad practice.  We have transform functions to do exactly these operations and do them efficiently.  It's never a good idea to have multiple copies of the same code repeated multiple time in the hope that "manually inlining" will make the code more efficient.  We should prefer calling functions for maintainability and readability until the code is proven to be a bottleneck and needs optimization.  Even then, we can use template functions to implement that as efficiently as this without keeping multiple copies of the code.